rust/agama-network/src/dbus/tree.rs

```diff
@@ -0,0 +1,229 @@
+                &path,
+                interfaces::Wireless::new(self.actions.clone(), Arc::clone(&cloned)),
+            )
+            .await?;
+        }
+
+        objects.register_connection(conn.uuid(), &path);
+        Ok(())
+    }
+
+    /// Removes a connection from the tree
+    ///
+    /// * `uuid`: UUID of the connection to remove.
+    pub async fn remove_connection(&mut self, uuid: Uuid) -> Result<(), ServiceError> {
+        let mut objects = self.objects.lock();
+        let Some(path) = objects.connection_path(uuid) else {
+            return Ok(())
+        };
+        self.remove_connection_on(path).await?;
+        objects.deregister_connection(uuid).unwrap();
+        Ok(())
+    }
+
+    /// Adds connections to the D-Bus tree.
+    ///
+    /// * `connections`: list of connections.
+    async fn add_connections(&self, connections: &Vec<Connection>) -> Result<(), ServiceError> {
+        for conn in connections.iter() {
+            self.add_connection(conn).await?;
+        }
+
+        self.add_interface(
+            CONNECTIONS_PATH,
+            interfaces::Connections::new(Arc::clone(&self.objects), self.actions.clone()),
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /// Clears all the connections from the tree.
+    async fn remove_connections(&self) -> Result<(), ServiceError> {
+        let mut objects = self.objects.lock();
+        for path in objects.connections.values() {
+            self.remove_connection_on(path.as_str()).await?;
+        }
+        objects.connections.clear();
+        Ok(())
+    }
+
+    /// Clears all the devices from the tree.
+    async fn remove_devices(&mut self) -> Result<(), ServiceError> {
+        let object_server = self.connection.object_server();
+        let mut objects = self.objects.lock();
+        for path in objects.devices.values() {
+            object_server
+                .remove::<interfaces::Device, _>(path.as_str())
+                .await?;
+        }
+        objects.devices.clear();
+        Ok(())
+    }
+
+    /// Removes a connection object on the given path
+    ///
+    /// * `path`: connection D-Bus path.
+    async fn remove_connection_on(&self, path: &str) -> Result<(), ServiceError> {
+        let object_server = self.connection.object_server();
+        _ = object_server.remove::<interfaces::Wireless, _>(path).await;
+        object_server.remove::<interfaces::Ipv4, _>(path).await?;
+        object_server
+            .remove::<interfaces::Connection, _>(path)
+            .await?;
+        Ok(())
+    }
+
+    async fn add_interface<T>(&self, path: &str, iface: T) -> Result<bool, ServiceError>
+    where
+        T: zbus::Interface,
+    {
+        let object_server = self.connection.object_server();
+        Ok(object_server.at(path.clone(), iface).await?)
+    }
+}
+
+/// Objects paths for known devices and connections
+#[derive(Debug, Default)]
+pub struct ObjectsRegistry {
+    pub devices: HashMap<String, String>,
```

Comment:
also, wouldn't zvariant::OwnedObjectPath be a more appropriate type?

Reply:
I have added a TODO to not forget about it.

Reply:
> you must also deal with invalid paths using that type.

Yes, but only once, which is the point. But it's fine to postpone it